src/components/TracePage/TraceTimelineViewer/TimelineHeaderRow.js

```diff
@@ -0,0 +1,190 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import * as React from 'react';
+import cx from 'classnames';
+
+import Ticks from './Ticks';
+import TimelineRow from './TimelineRow';
+
+import './TimelineHeaderRow.css';
+
+type TimelineHeaderRowProps = {
+  endTime: number,
+  nameColumnWidth: number,
+  numTicks: number,
+  onColummWidthChange: number => void,
+  startTime: number,
+};
+
+type TimelineColumnResizerProps = {
+  min: number,
+  max: number,
+  onChange: number => void,
+  position: number,
+};
+
+type TimelineColumnResizerState = {
+  currentDragPosition: ?number,
+  rootX: ?number,
+};
+
+const LEFT_MOUSE_BUTTON = 0;
+
+class TimelineColumnResizer extends React.PureComponent<
+  TimelineColumnResizerProps,
+  TimelineColumnResizerState
+> {
+  props: TimelineColumnResizerProps;
+  state: TimelineColumnResizerState;
+  _rootElm: ?Element;
+  _isDragging: boolean;
+
+  constructor(props) {
+    super(props);
+    this._rootElm = undefined;
+    this._isDragging = false;
+    this.state = {
+      currentDragPosition: null,
+      rootX: null,
+    };
+    this._setRootElm = this._setRootElm.bind(this);
+    this._onDraggerMouseDown = this._onDraggerMouseDown.bind(this);
+    this._onWindowMouseMove = this._onWindowMouseMove.bind(this);
+    this._onWindowMouseUp = this._onWindowMouseUp.bind(this);
+  }
+
+  componentWillUnmount() {
+    if (this._isDragging) {
+      window.removeEventListener('mousemove', this._onWindowMouseMove);
+      window.removeEventListener('mouseup', this._onWindowMouseUp);
+      this._isDragging = false;
+    }
+  }
+
+  _setRootElm = function _setRootElm(elm) {
+    this._rootElm = elm;
+  };
+
+  _onDraggerMouseDown = function _onDraggerMouseDown({ button, clientX }) {
+    if (this._isDragging || button !== LEFT_MOUSE_BUTTON || !this._rootElm) {
+      return;
+    }
+    const rootX = this._rootElm.getBoundingClientRect().left;
+    this.setState({
+      rootX,
+      currentDragPosition: clientX - rootX,
+    });
+    window.addEventListener('mousemove', this._onWindowMouseMove);
+    window.addEventListener('mouseup', this._onWindowMouseUp);
+    this._isDragging = true;
+    if (document && document.body && document.body.style) {
+      (document.body.style: any).userSelect = 'none';
+    }
+  };
+
+  _onWindowMouseMove = function _onWindowMouseMove({ clientX }) {
+    const { rootX } = this.state;
+    this.setState({ rootX, currentDragPosition: clientX - (rootX || 0) });
+  };
+
+  _onWindowMouseUp = function _onWindowMouseUp({ clientX }) {
```

Comment:
We seem to be doing a lot of stuff with dragging and complex calculations inline react components that can be hard to follow in future development.

Is there a solid library we can use that abstracts this stuff out or an HOC pattern we can reuse here?

Reply:
Went ahead and extracted the calculations out into their [own function](https://github.com/uber/jaeger-ui/pull/74/commits/00e83d37a3f558f931415725783a8267c7e688ec#diff-7c366cc806a433631c209ceb514c9877R89) and [simplified](https://github.com/uber/jaeger-ui/pull/74/commits/00e83d37a3f558f931415725783a8267c7e688ec#diff-7c366cc806a433631c209ceb514c9877R154) the way the dimensions of the highlight are calculated (via using `left`, `right` instead of `width`—simpler calculations).

If you think it needs more changes, can you elaborate?